File: run_trend_detector_and_plot.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include <math.h>
#include <errno.h>
#include <tensorflow/lite/c/c_api.h>

#define COLOR_RED   0xFF0000
#define COLOR_BLUE  0x0000FF
#define COLOR_GREY  0x808080

typedef struct {
    float time;
    float temp;
} DataPoint;

typedef struct {
    int heating;
    int cooling;
    int constant;
} TrendCounts;

static void die(const char *msg) {
    perror(msg);
    exit(EXIT_FAILURE);
}

static int cmp_time(const void *a, const void *b) {
    const DataPoint *da = (const DataPoint *)a;
    const DataPoint *db = (const DataPoint *)b;
    if (da->time < db->time) return -1;
    if (da->time > db->time) return 1;
    return 0;
}

static DataPoint *load_csv(const char *path, size_t *out_len) {
    FILE *fp = fopen(path, "r");
    if (!fp) die("fopen CSV");

    char line[256];
    if (!fgets(line, sizeof(line), fp)) die("read header");

    size_t cap = 1024, len = 0;
    DataPoint *arr = malloc(cap * sizeof(DataPoint));
    if (!arr) die("malloc");
    while (fgets(line, sizeof(line), fp)) {
        char *tok = strtok(line, ",");
        if (!tok) continue;
        float rel_time = strtof(tok, NULL);
        tok = strtok(NULL, ",");
        if (!tok) continue;
        float temperature = strtof(tok, NULL);

        if (len == cap) {
            cap *= 2;
            arr = realloc(arr, cap * sizeof(DataPoint));
            if (!arr) die("realloc");
        }
        arr[len].time = rel_time;
        arr[len].temp = temperature;
        len++;
    }

    fclose(fp);
    qsort(arr, len, sizeof(DataPoint), cmp_time);
    *out_len = len;
    return arr;
}

static TfLiteInterpreter *init_interpreter(const char *model_path) {
    TfLiteModel *model = TfLiteModelCreateFromFile(model_path);
    if (!model) {
        fprintf(stderr, "Failed to load model: %s\n", model_path);
        exit(EXIT_FAILURE);
    }
    TfLiteInterpreterOptions *options = TfLiteInterpreterOptionsCreate();
    TfLiteInterpreterOptionsSetNumThreads(options, 1);

    TfLiteInterpreter *interpreter = TfLiteInterpreterCreate(model, options);
    TfLiteInterpreterAllocateTensors(interpreter);

    /* Print I/O details */
    int input_cnt = TfLiteInterpreterGetInputTensorCount(interpreter);
    int output_cnt = TfLiteInterpreterGetOutputTensorCount(interpreter);
    printf("Model input tensors: %d, output tensors: %d\n", input_cnt, output_cnt);
    for (int i = 0; i < input_cnt; ++i) {
        const TfLiteTensor *t = TfLiteInterpreterGetInputTensor(interpreter, i);
        printf("  Input %d: type=%d dims=" , i, TfLiteTensorType(t));
        int dims = TfLiteTensorNumDims(t);
        for (int d = 0; d < dims; ++d) {
            printf("%s%d", d==0?"[":" ", TfLiteTensorDim(t,d));
        }
        printf("]\n");
    }
    for (int i = 0; i < output_cnt; ++i) {
        const TfLiteTensor *t = TfLiteInterpreterGetOutputTensor(interpreter, i);
        printf("  Output %d: type=%d dims=" , i, TfLiteTensorType(t));
        int dims = TfLiteTensorNumDims(t);
        for (int d = 0; d < dims; ++d) {
            printf("%s%d", d==0?"[":" ", TfLiteTensorDim(t,d));
        }
        printf("]\n");
    }

    TfLiteModelDelete(model);   /* interpreter holds reference */
    TfLiteInterpreterOptionsDelete(options);
    return interpreter;
}

static int predict_trend(TfLiteInterpreter *interp,
                         const float *temps_window,
                         const float *times_window,
                         int window) {
    int input_cnt = TfLiteInterpreterGetInputTensorCount(interp);
    if (input_cnt != 2) {
        fprintf(stderr, "Model expected 2 input tensors but got %d\n", input_cnt);
        exit(EXIT_FAILURE);
    }

    /* Ensure input tensor dims match window length; resize if needed */
    for (int idx = 0; idx < 2; ++idx) {
        TfLiteTensor *tensor = TfLiteInterpreterGetInputTensor(interp, idx);
        int ndims = TfLiteTensorNumDims(tensor);
        bool needs_resize = true;
        if (ndims == 1 && TfLiteTensorDim(tensor,0) == window) needs_resize = false;
        if (needs_resize) {
            int new_dims[1] = {window};
            TfLiteInterpreterResizeInputTensor(interp, idx, new_dims, 1);
        }
    }
    TfLiteInterpreterAllocateTensors(interp);

    /* Copy data into tensors */
    TfLiteTensor *t0 = TfLiteInterpreterGetInputTensor(interp, 0);
    TfLiteTensorCopyFromBuffer(t0, temps_window, window*sizeof(float));
    TfLiteTensor *t1 = TfLiteInterpreterGetInputTensor(interp, 1);
    TfLiteTensorCopyFromBuffer(t1, times_window, window*sizeof(float));

    /* Invoke */
    if (TfLiteInterpreterInvoke(interp) != kTfLiteOk) {
        fprintf(stderr, "Interpreter invoke failed\n");
        exit(EXIT_FAILURE);
    }

    const TfLiteTensor *out = TfLiteInterpreterGetOutputTensor(interp, 0);
    int32_t val;
    TfLiteTensorCopyToBuffer(out, &val, sizeof(int32_t));
    return (int)val;
}

/*-------------------------------------------------------------
 * MAIN PIPELINE
 *-----------------------------------------------------------*/

static void write_plot_files(const DataPoint *dp, const int *pred, size_t n) {
    FILE *pf = fopen("points.dat", "w");
    if (!pf) die("points.dat");
    for (size_t i = 0; i < n; ++i) {
        uint32_t color = (pred[i]==1) ? COLOR_RED : (pred[i]==-1 ? COLOR_BLUE : COLOR_GREY);
        fprintf(pf, "%f %f 0x%06x\n", dp[i].time, dp[i].temp, color);
    }
    fclose(pf);

    FILE *gp = fopen("plot.gp", "w");
    if (!gp) die("plot.gp");
    fprintf(gp,
        "set terminal pngcairo size 1400,600\n"
        "set output 'plot.png'\n"
        "set title 'Temperature Trend Detection – Red: Heating, Blue: Cooling, Grey: Constant/Fluctuating'\n"
        "set xlabel 'Relative Time'\n"
        "set ylabel 'Temperature (°C)'\n"
        "set grid\n"
        "plot 'points.dat' using 1:2:3 with points pt 7 ps 0.5 lc rgb var notitle\n");
    fclose(gp);
    system("gnuplot plot.gp");
}

int main(int argc, char **argv) {
    /* Default args */
    const char *csv_path = "filtered_temperature_data_2.csv";
    const char *model_path = "trend_predictor.tflite";
    int window = 8;

    /* Simple arg parsing */
    for (int i = 1; i < argc; ++i) {
        if (!strcmp(argv[i], "--csv") && i+1 < argc) {
            csv_path = argv[++i];
        } else if (!strcmp(argv[i], "--model") && i+1 < argc) {
            model_path = argv[++i];
        } else if (!strcmp(argv[i], "--window") && i+1 < argc) {
            window = atoi(argv[++i]);
        } else {
            fprintf(stderr, "Usage: %s [--csv PATH] [--model PATH] [--window N]\n", argv[0]);
            return EXIT_FAILURE;
        }
    }

    size_t n_points = 0;
    DataPoint *data = load_csv(csv_path, &n_points);
    printf("Loaded %zu data points from %s\n", n_points, csv_path);

    TfLiteInterpreter *interp = init_interpreter(model_path);

    if (window <= 0) {
        const TfLiteTensor *in0 = TfLiteInterpreterGetInputTensor(interp, 0);
        if (TfLiteTensorNumDims(in0) >= 1) {
            window = TfLiteTensorDim(in0, 0);
        } else {
            window = 1;
        }
    }
    printf("Using window size: %d\n", window);

    float *temp_window = malloc(window * sizeof(float));
    float *time_window = malloc(window * sizeof(float));
    int *predictions = malloc(n_points * sizeof(int));

    TrendCounts counts = {0,0,0};

    for (size_t idx = 0; idx < n_points; ++idx) {
        int start = (int)idx - window + 1;
        if (start < 0) start = 0;
        int len = (int)idx - start + 1;
        int pad = window - len;
        /* Pad with first value in window */
        for (int i = 0; i < pad; ++i) {
            temp_window[i] = data[start].temp;
            time_window[i] = data[start].time;
        }
        for (int i = 0; i < len; ++i) {
            temp_window[pad + i] = data[start + i].temp;
            time_window[pad + i] = data[start + i].time;
        }

        int pred = predict_trend(interp, temp_window, time_window, window);
        predictions[idx] = pred;
        if (pred == 1) counts.heating++;
        else if (pred == -1) counts.cooling++;
        else counts.constant++;
    }

    printf("Trend counts: {1: %d, -1: %d, 0: %d}\n", counts.heating, counts.cooling, counts.constant);

    write_plot_files(data, predictions, n_points);

    TfLiteInterpreterDelete(interp);
    free(data);
    free(temp_window);
    free(time_window);
    free(predictions);

    return 0;
} 

File: run_trend_detector_and_plot.py
import argparse
from pathlib import Path
from typing import List

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

try:
    from tflite_runtime.interpreter import Interpreter  # type: ignore
except ImportError:
    from tensorflow.lite.python.interpreter import Interpreter  # type: ignore


def load_data(csv_path: Path) -> pd.DataFrame:
    """Load the CSV containing `relative_time` and `temperature` columns."""
    df = pd.read_csv(csv_path)
    expected_cols = {"relative_time", "temperature"}
    if not expected_cols.issubset(df.columns):
        raise ValueError(
            f"CSV must contain columns {expected_cols}, found {set(df.columns)}"
        )
    df = df.sort_values("relative_time").reset_index(drop=True)
    return df


def init_interpreter(model_path: Path) -> Interpreter:
    """Initialise a TensorFlow Lite interpreter from a model file."""
    interpreter = Interpreter(model_path=str(model_path))
    interpreter.allocate_tensors()

    input_details = interpreter.get_input_details()
    output_details = interpreter.get_output_details()
    print("Model input details:")
    for d in input_details:
        print(d)
    print("\nModel output details:")
    for d in output_details:
        print(d)

    return interpreter


def predict_trend(
    interpreter: Interpreter,
    temps_window: np.ndarray,
    times_window: np.ndarray,
    last_trend: int,
) -> int:
    """Run inference on a triple of inputs (temps, times & last trend).

    The updated TFLite model now expects three inputs:
        0 → temperatures   – 1-D window of floats of length *window*
        1 → times          – 1-D window of floats of length *window*
        2 → last_trend     – scalar `int32` carrying the previously predicted trend

    The output is a scalar `int32` with values:
        1  – heating   | -1 – cooling   | 0 – constant / fluctuating
    """

    input_details = interpreter.get_input_details()
    if len(input_details) != 3:
        raise RuntimeError(
            f"Expected the model to have 3 inputs (temps, times, last_trend) but got {len(input_details)}"
        )

    # Prepare the three input arrays in the same order as the model definition.
    inputs = [
        temps_window.astype(np.float32),
        times_window.astype(np.float32),
        np.array([last_trend], dtype=np.int32),
    ]

    # Iterate through each tensor, resizing if necessary (dynamic shapes) and feeding data.
    for idx, arr in enumerate(inputs):
        detail = input_details[idx]
        required_shape = tuple(detail["shape"])

        # If the interpreter uses *None* as a placeholder for dynamic dimension, treat that as compatible size.
        compatible = (
            required_shape == arr.shape
            or any(dim is None for dim in required_shape)
        )

        if not compatible:
            try:
                interpreter.resize_tensor_input(detail["index"], arr.shape)
                interpreter.allocate_tensors()
                # refresh details after allocation
                detail = interpreter.get_input_details()[idx]
            except Exception:
                try:
                    arr = arr.reshape(detail["shape"])
                except ValueError as exc:
                    raise ValueError(
                        f"Input {idx} array has shape {arr.shape} which cannot be reshaped to {detail['shape']}"
                    ) from exc

        # Ensure dtype matches
        arr = arr.astype(detail["dtype"])
        interpreter.set_tensor(detail["index"], arr)

    interpreter.invoke()

    output_detail = interpreter.get_output_details()[0]
    output_val = interpreter.get_tensor(output_detail["index"]).squeeze()

    return int(output_val)


def run(csv_path: Path, model_path: Path, window: int = 8):
    df = load_data(csv_path)
    interpreter = init_interpreter(model_path)

    if window <= 0:
        input_shape = interpreter.get_input_details()[0]["shape"]
        if len(input_shape) >= 2:
            window = int(input_shape[1])
        else:
            window = 1

    temps = df["temperature"].values.astype(np.float32)
    times = df["relative_time"].values.astype(np.float32)

    predictions: List[int] = []
    prev_trend: int = 0  # Unknown at start
    for idx in range(len(temps)):
        start = max(0, idx - window + 1)
        temp_window = temps[start : idx + 1]
        time_window = times[start : idx + 1]

        if len(temp_window) < window:
            pad_size = window - len(temp_window)
            pad_temp = np.full(pad_size, temp_window[0], dtype=np.float32)
            pad_time = np.full(pad_size, time_window[0], dtype=np.float32)
            temp_window = np.concatenate([pad_temp, temp_window])
            time_window = np.concatenate([pad_time, time_window])

        pred = predict_trend(interpreter, temp_window, time_window, prev_trend)
        predictions.append(pred)
        prev_trend = pred  # feed the latest prediction into the next iteration

    predictions = np.array(predictions)

    unique, counts = np.unique(predictions, return_counts=True)
    print("Trend counts:", dict(zip(unique, counts)))

    color_map = {-1: "blue", 1: "red", 0: "grey"}
    cmap = [color_map.get(p, "black") for p in predictions]

    plt.figure(figsize=(14, 6))
    plt.scatter(times, temps, c=cmap, s=5)
    plt.title("Temperature Trend Detection – Red: Heating, Blue: Cooling, Grey: Constant/Fluctuating")
    plt.xlabel("Relative Time")
    plt.ylabel("Temperature (°C)")
    plt.grid(True)
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Run trend detection on temperature data and plot results.")
    parser.add_argument(
        "--csv",
        type=Path,
        default="filtered_temperature_data_2.csv",
        help="Path to the CSV file containing temperature data.",
    )
    parser.add_argument(
        "--model",
        type=Path,
        default="trend_predictor.tflite",
        help="Path to the TFLite model file.",
    )
    parser.add_argument(
        "--window",
        type=int,
        default=8,
        help="Number of consecutive samples to feed the model (sliding window).",
    )

    args = parser.parse_args()
    run(args.csv, args.model, args.window) 